### PR TITLE
Update `Join` documentation

### PR DESCRIPTION
**Pull request by matheus-consoli**
I've been thinking about working a bit on the public documentation of this crate.

My plan is to first revisit the traits in the next feel days, and then improve the first page/crate documentation and readme. Instead of opening a big-ass PR, I want to iterate over the traits one PR at a time, so we can have a nicer feedback loop.

This first PR should improve a bit the docs of `Join` (it also provides a few minor fixes on `lib.rs`).

Let me know what you think!

### CODE REVIEW

**Comment by matheus-consoli** [comment]
CI failed on fmting an empty line - I force-pushed the fix.

File: src/future/join/mod.rs [hunk 1]
```diff
@@ -7,19 +7,60 @@ pub(crate) mod vec;
 /// Wait for all futures to complete.
 ///
 /// Awaits multiple futures simultaneously, returning the output of the futures
-/// once all complete.
+/// in the same container type they we're created once all complete.
```
Comment (yoshuawuyts) [inline src/future/join/mod.rs:10]:
```suggestion
/// in the same container type they were created once all complete.
```

File: src/future/join/mod.rs [hunk 2]
```diff
@@ -7,19 +7,60 @@ pub(crate) mod vec;
...
     type Future: Future<Output = Self::Output>;
 
     /// Waits for multiple futures to complete.
     ///
-    /// Awaits multiple futures simultaneously, returning the output of the
-    /// futures once all complete.
+    /// Awaits multiple futures simultaneously, returning the output of the futures
+    /// in the same container type they we're created once all complete.
     ///
+    /// # Examples
```
Comment (yoshuawuyts) [inline src/future/join/mod.rs:23]:
```suggestion
    /// # Examples
    ///
```

File: src/future/join/mod.rs [hunk 3]
```diff
@@ -7,19 +7,60 @@ pub(crate) mod vec;
...
+    /// Awaiting multiple futures of the same type can be done using either a vector
+    /// or an array.
+    /// ```rust
+    /// #  futures::executor::block_on(async {
+    /// use futures_concurrency::prelude::*;
+    ///
+    /// // these can be whatever kind of `Future` you have,
+    /// // as long as it all have the exactly same type
```
Comment (yoshuawuyts) [inline src/future/join/mod.rs:31]:
```suggestion
    /// // all futures passed here are of the same type
```

File: src/future/join/mod.rs [hunk 4]
```diff
@@ -7,19 +7,60 @@ pub(crate) mod vec;
...
+    ///
+    /// let outputs = [fut1, fut2, fut3].join().await;
+    /// assert_eq!(outputs, [1, 2, 3]);
+    /// # })
+    /// ```
+    ///
+    /// But in real world, you might want to wait multiple futures of
+    /// different types - let's say you're awaiting on custom `async {}` blocks,
+    /// or that your futures have different `Output` types - for that, you can
+    /// `join` on tuples.
```
Comment (yoshuawuyts) [inline src/future/join/mod.rs:44]:
```suggestion
    /// In practice however, it's common to want to await multiple futures of
    /// different types. For example if you have two different `async {}` blocks,
    /// you want to `.await`. To do that, you can call `.join` on tuples of futures.
```

File: src/future/join/mod.rs [hunk 5]
```diff
@@ -7,19 +7,60 @@ pub(crate) mod vec;
...
+    /// ```rust
+    /// #  futures::executor::block_on(async {
+    /// use futures_concurrency::prelude::*;
+    ///
+    /// async fn some_async_fn() -> usize { 3 }
+    ///
+    /// // these can be whatever kind of `Future` you have,
+    /// // even those that have totally different concrete types.
```
Comment (yoshuawuyts) [inline src/future/join/mod.rs:52]:
```suggestion
    /// // the futures passed here are of different types
```

**Review by yoshuawuyts — APPROVED** [review]
Thanks for submitting this. I appreciate you working to improve the docs! I've added some comments. If we merge those then I think this PR should be good to go!

**Pull request merged**